example of splitting up an FX graph into smaller subgraphs with own submodules

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45404 example of splitting up an FX graph into smaller subgraphs with own submodules**

Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

Commit:
example of splitting up an FX graph into smaller subgraphs with own submodules

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Lilyjjo commented on torch/fx/experimental/subgraph_creation_example.py:
```diff
@@ -0,0 +1,211 @@
+                partition.targets[node.target] = target_attr
+                target = target_atoms[-1]
+
+            new_node = partition.graph.create_node(op=node.op, target=target, args=gathered_args,
```
mypy error: Argument "args" to "create_node" of "Graph" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Optional[Tuple[Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None], ...]]"  [arg-type]

Lilyjjo commented on torch/fx/experimental/subgraph_creation_example.py:
```diff
@@ -0,0 +1,211 @@
+        base_mod_attrs[submod_name] = GraphModule(partition.targets, partition.graph)
+
+        # Emit call in base graph to this submodule
+        output_val = base_mod_graph.call_module(submod_name, [base_mod_env[name] for name in partition.inputs])
```
There's a similar mypy error here also due to the type of the args... 

error: Argument 2 to "call_module" of "Graph" has incompatible type "List[Node]"; expected "Optional[Tuple[Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None], ...]]"  [arg-type]

Lilyjjo commented on torch/fx/experimental/subgraph_creation_example.py:
```diff
@@ -0,0 +1,211 @@
+            # Unpack multiple return values from submodule
+            output_val_proxy = torch.fx.proxy.Proxy(output_val)
+            for i, output_name in enumerate(partition.outputs):
+                base_mod_env[output_name] = output_val_proxy[i].node
```
The last mypy error isn't realted to the first three, the Proxy class is lacking an indexing method, not sure if I should add one or if there is another workaround?

mypy error: Value of type "Proxy" is not indexable  [index]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Lilyjjo commented on torch/fx/experimental/subgraph_creation_example.py:
```diff
@@ -0,0 +1,212 @@
+                partition.targets[node.target] = target_attr
+                target = target_atoms[-1]
+
+            new_node = partition.graph.create_node(op=node.op, target=target, args=gathered_args,  # type: ignore 
```
There are two mypy errors here that I'm not sure how to go about resolving, both are related to the types of the args/kwargs

mypy error: Argument "args" to "create_node" of "Graph" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Optional[Tuple[Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None], ...]]" [arg-type]

error: Argument "kwargs" to "create_node" of "Graph" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Optional[Dict[str, Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]]]" [arg-type]

jamesr66a replied:
I can't repro this locally but I think if you put:
```
assert isinstance(gathered_args, tuple)
assert isinstance(gathered_kwargs, dict)
```

Before this line, mypy might match the types correctly

Lilyjjo replied:
yep this fixed it-- thanks

Lilyjjo commented on torch/fx/experimental/subgraph_creation_example.py:
```diff
@@ -0,0 +1,212 @@
+            # Unpack multiple return values from submodule
+            output_val_proxy = torch.fx.proxy.Proxy(output_val)
+            for i, output_name in enumerate(partition.outputs):
+                base_mod_env[output_name] = output_val_proxy[i].node  # type: ignore
```
The last mypy error isn't realted to the first three, the Proxy class doesn't like to be indexed

mypy error: Value of type "Proxy" is not indexable [index]

jamesr66a replied:
See if this works?

https://gist.github.com/jamesr66a/642a92a473936487cb33bd424a45f35a

Otherwise I'm not sure, and it's probably okay to keep the ignore annotation

Lilyjjo replied:
Yeah it's the same issue with the changes to proxy, I kept the ignore annotation

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]

Commit:
Update on "example of splitting up an FX graph into smaller subgraphs with own submodules"


Differential Revision: [D23956147](https://our.internmc.facebook.com/intern/diff/D23956147)

[ghstack-poisoned]
